File: leetcode/3/main.cpp
#include <algorithm>
#include <string>
#include <unordered_map>
using namespace std;

template <typename T, typename V> using U_Map = unordered_map<T, V>;

class Solution {
public:
  int lengthOfLongestSubstring(string s) {
    const int N = s.length();
    if (N == 0) return 0;

    U_Map<char, int> m;
    m.insert({ s[0], 0 });

    int left = 0;
    int answer = 1;
    for (int i = 1; i < N; i++) {
      if (m.contains(s[i])) { left = max(left, m[s[i]] + 1); }
      m[s[i]] = i;

      answer = max(answer, i - left + 1);
    }

    return answer;
  }
};

File: leetcode/3/main.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
  const n = s.length;
  if (n == 0) return 0;

  const m = new Map();
  m.set(s[0], 0);

  let answer = 1,
    left = 0;

  for (let i = 1; i < n; i++) {
    if (m.has(s[i])) left = Math.max(left, m.get(s[i]) + 1);
    m.set(s[i], i);

    answer = Math.max(answer, i - left + 1);
  }

  return answer;
};
